src/ring_theory/perfection.lean

```diff
@@ -113,7 +115,106 @@ lemma coeff_ne_zero_of_le {f : ring.perfection R p} {m n : ℕ} (hfm : coeff R p
+      f.map_add _ _ },
+  inv_fun := ring_hom.comp $ coeff S p 0,
+  left_inv := λ f, ring_hom.ext $ λ r, rfl,
+  right_inv := λ f, ring_hom.ext $ λ r, ext $ λ n,
+    show coeff S p 0 (f (_root_.pth_root R p ^[n] r)) = coeff S p n (f r),
+    by rw [← coeff_iterate_frobenius _ 0 n, zero_add, ← ring_hom.map_iterate_frobenius,
+      right_inverse_pth_root_frobenius.iterate] }
+
+end perfection
+
+/-- A perfection map to a ring of characteristic `p` is a map that is isomorphic
+to its perfection. -/
+@[nolint has_inhabited_instance] structure perfection_map (p : ℕ) [fact p.prime]
+  {R : Type u₁} [comm_semiring R] [char_p R p]
+  {P : Type u₂} [comm_semiring P] [char_p P p] [perfect_ring P p] (π : P →+* R) : Prop :=
+(injective : ∀ ⦃x y : P⦄, (∀ n, π (pth_root P p ^[n] x) = π (pth_root P p ^[n] y)) → x = y)
+(surjective : ∀ f : ℕ → R, (∀ n, f (n + 1) ^ p = f n) →
+  ∃ x : P, ∀ n, π (pth_root P p ^[n] x) = f n)
+
+namespace perfection_map
+
+variables {p : ℕ} [fact p.prime]
+variables {R : Type u₁} [comm_semiring R] [char_p R p]
+variables {P : Type u₂} [comm_semiring P] [char_p P p] [perfect_ring P p]
+
+/-- Create a `perfection_map` from an isomorphism to the perfection. -/
+@[simps] lemma mk' {f : P →+* R} (g : P ≃+* ring.perfection R p)
+  (hfg : perfection.extend p P R f = g) :
+  perfection_map p f :=
+{ injective := λ x y hxy, g.injective $ (ring_hom.ext_iff.1 hfg x).symm.trans $
+    eq.symm $ (ring_hom.ext_iff.1 hfg y).symm.trans $ perfection.ext $ λ n, (hxy n).symm,
+  surjective := λ y hy, let ⟨x, hx⟩ := g.surjective ⟨y, hy⟩ in
+    ⟨x, λ n, show perfection.coeff R p n (perfection.extend p P R f x) =
+        perfection.coeff R p n ⟨y, hy⟩,
+      by rw [hfg, ← coe_fn_coe_base, hx]⟩ }
+
+variables (p R P)
+
+/-- The canonical perfection map from the perfection of a ring. -/
+lemma of : perfection_map p (perfection.coeff R p 0) :=
+mk' (ring_equiv.refl _) $ (equiv.apply_eq_iff_eq_symm_apply _).2 rfl
+
+/-- For a perfect ring, it itself is the perfection. -/
+def id [perfect_ring R p] : perfection_map p (ring_hom.id R) :=
+{ injective := λ x y hxy, hxy 0,
+  surjective := λ f hf, ⟨f 0, λ n, show pth_root R p ^[n] (f 0) = f n,
+    from nat.rec_on n rfl $ λ n ih, injective_pow_p p $
+      by rw [function.iterate_succ_apply', pth_root_pow_p _, ih, hf]⟩ }
+
+variables {p R P}
+/-- A perfection map induces an isomorphism to the prefection. -/
+noncomputable def equiv {π : P →+* R} (m : perfection_map p π) : P ≃+* ring.perfection R p :=
+ring_equiv.of_bijective (perfection.extend p P R π)
+⟨λ x y hxy, m.injective $ λ n, (congr_arg (perfection.coeff R p n) hxy : _),
+λ f, let ⟨x, hx⟩ := m.surjective f.1 f.2 in ⟨x, perfection.ext $ hx⟩⟩
+
+variables (p R P)
+/-- Given rings `R` and `S` of characteristic `p`, with `R` being perfect,
+any homomorphism `R →+* S` can be extended to a homomorphism `R →+* P`,
+where `P` is any perfection of `S`. -/
+@[simps] noncomputable def extend [perfect_ring R p] (S : Type u₂) [comm_semiring S] [char_p S p]
```

Comment:
Should this be called `lift`?

Reply:
I don't know any more. its neighbour `(K →+* L) ≃ (perfect_closure K p →+* L)` is called `lift`, so I figured this one should be called `extend` just for the sake of it:
https://github.com/leanprover-community/mathlib/blob/39ecd1abaabede897b2e5367efc2af8cdcadcc5b/src/field_theory/perfect_closure.lean#L382-L383

Reply:
Hmm, I would rather call `perfect_closure.lift` an example of extending something...

Reply:
Yeah so I figured that we do a switcheroo in Leanspeak

Reply:
Like Greenland and Iceland.